### PR TITLE
[Container scalability] Implementation of backup file logic for MySqlAccountService and addition of LMT field to Accounts/Containers

### PR DESCRIPTION
**Pull request by Arun-LinkedIn**
Changes to:
1. Persist account and container metadata to local file on disk after syncing recent updates from db.   
2. Retrieve metadata from local file on start up and initialize cache.
3. Add 'lastModifiedTime' field to Account and Container class, update it from timestamp column in DB records and use the max LMT as reference timestamp during next db sync.

### CODE REVIEW

File: ambry-account/src/main/java/com/github/ambry/account/MySqlAccountService.java [hunk 1]
```diff
@@ -307,4 +342,126 @@ private void updateContainersWithMySqlStore(short accountId, Collection<Containe
       }
     }
   }
+
+  public static class MySqlBackupFileManager {
+    private static final Logger logger = LoggerFactory.getLogger(BackupFileManager.class);
```
Comment (lightningrob) [inline ambry-account/src/main/java/com/github/ambry/account/MySqlAccountService.java:347]: MySqlBackupFileManager.  Also, this class is big enough to be in its own file.

File: ambry-account/src/main/java/com/github/ambry/account/MySqlAccountService.java [hunk 2]
```diff
@@ -307,4 +359,134 @@ private void updateContainersWithMySqlStore(short accountId, Collection<Containe
...
+    private static final String BACKUP_FILE_NAME = "MySqlAccountMetadataCopy";
+    private final AccountServiceMetrics accountServiceMetrics;
+    static final String TEMP_FILE_SUFFIX = "tmp";
+    static final String SEP = ".";
+    static final DateTimeFormatter TIMESTAMP_FORMATTER = DateTimeFormatter.ofPattern("yyyyMMdd'T'HHmmss");
+    static final ZoneOffset zoneOffset = ZoneId.systemDefault().getRules().getOffset(LocalDateTime.now());
+    static final Pattern filenamePattern = Pattern.compile(BACKUP_FILE_NAME + "\\.(\\d{8}T\\d{6})$");
```
Comment (justinlin-linkedin) [inline ambry-account/src/main/java/com/github/ambry/account/MySqlAccountService.java:371]: you might want to also add a begin anchor of "^" for your pattern. Something like
`"^" + BACKUP_FILE_NAME + "......"`

**Review by justinlin-linkedin — COMMENTED** [review]
It looks pretty good to me, please add more unit tests to cover your new backup file implementation.

**Review by lightningrob — CHANGES_REQUESTED** [review]

File: ambry-api/src/main/java/com/github/ambry/config/MySqlAccountServiceConfig.java [hunk 3]
```diff
@@ -73,6 +74,14 @@
   @Default("")
   public final String backupDir;
 
+  /**
+   * The maximum number of local backup files kept in disk. When account service exceeds this count, every time it creates
+   * a new backup file, it will remove the oldest one.
+   */
+  @Config(MAX_BACKUP_FILE_COUNT)
+  @Default("100")
```
Comment (jsjtzyy) [inline ambry-api/src/main/java/com/github/ambry/config/MySqlAccountServiceConfig.java:82]: minor: I guess you wanted to set 10 here.

File: ambry-account/src/test/java/com/github/ambry/account/BackupFileManagerTest.java [hunk 4]
```diff
@@ -16,6 +16,7 @@
 import com.codahale.metrics.MetricRegistry;
 import com.github.ambry.config.HelixAccountServiceConfig;
 import com.github.ambry.config.VerifiableProperties;
+import com.github.ambry.utils.SystemTime;
```
Comment (jsjtzyy) [inline ambry-account/src/test/java/com/github/ambry/account/BackupFileManagerTest.java:19]: This import can be removed.

**Review by jsjtzyy — APPROVED** [review]

**Review by lightningrob — APPROVED** [review]
Looks good, especially the refactoring of BackupFileManager to make it generic and reusable.
Will merge once the build is green.

File: ambry-api/src/test/java/com/github/ambry/account/AccountContainerTest.java [hunk 5]
```diff
@@ -1041,6 +1044,7 @@ private JSONObject buildContainerJson(Container container) throws JSONException
         containerJson.put(DESCRIPTION_KEY, container.getDescription());
         containerJson.put(IS_PRIVATE_KEY, !container.isCacheable());
         containerJson.put(PARENT_ACCOUNT_ID_KEY, container.getParentAccountId());
+        containerJson.put(LAST_MODIFIED_TIME_KEY, container.getLastModifiedTime());
```
Comment (lightningrob) [inline ambry-api/src/test/java/com/github/ambry/account/AccountContainerTest.java:1047]: Is this needed in the old version?

File: ambry-api/src/test/java/com/github/ambry/account/AccountContainerTest.java [hunk 6]
```diff
@@ -34,6 +35,7 @@
 
 import static com.github.ambry.account.Account.*;
 import static com.github.ambry.account.Container.*;
+import static com.github.ambry.account.Container.LAST_MODIFIED_TIME_KEY;
```
Comment (lightningrob) [inline ambry-api/src/test/java/com/github/ambry/account/AccountContainerTest.java:38]: Not sure this was really needed.  Could have referenced it by class.

**Pull request merged**